peps/pep-0749.rst

```diff
@@ -0,0 +1,741 @@
+is exclusively meant for type hints.
+
+*Add the functionality to the types module*: The :py:mod:`types` module is
+meant for functionality related to *types*, and annotations can exist on functions
+and modules, not only on types.
+
+*Develop this functionality in a third-party package*: The functionality in this new
+module will be pure Python code, and it is possible to implement a third-party package
+that provides the same functionality by interacting directly with ``__annotate__``
+functions generated by the interpreter. However, the functionality of the proposed new
+module will certainly be useful in the standard library itself (e.g., for implementing
+:py:mod:`dataclasses` and :py:class:`typing.NamedTuple`), so it makes sense to include
+it in the standard library.
+
+*Add this functionality to a private module*: It would be possible to initially develop
+the module in a private standard library module (e.g., ``_annotations``), and publicize
+it only after we have gained more experience with the API. However, we already know
+that we will need parts of this module for the standard library itself (e.g., for
+implementing :py:mod:`!dataclasses` and :py:class:`!typing.NamedTuple`). Even if we make
+it private, the module will inevitably get used by third-party users. It is preferable
+to start with a clear, documented API from the beginning, to enable third-party users to
+support :pep:`649` semantics as thoroughly as the standard library. The module will
+immediately be used in other parts of the standard library, ensuring that it covers a
+reasonable set of use cases.
+
+Behavior of the REPL
+====================
+
+:pep:`649` specifies the following behavior of the interactive REPL:
+
+    For the sake of simplicity, in this case we forego delayed evaluation.
+    Module-level annotations in the REPL shell will continue to work exactly
+    as they do with “stock semantics”, evaluating immediately and setting the
+    result directly inside the ``__annotations__`` dict.
+
+There are several problems with this proposed behavior. It makes the REPL the
+only context where annotations are still evaluated immediately, which is
+confusing for users and complicates the language.
+
+It also makes the implementation of the REPL more complex, as it needs to
+ensure that all statements are compiled in "interactive" mode, even if their
+output does not need to be displayed. (This matters if there are multiple
+statements in a single line evaluated by the REPL.)
+
+Most importantly, this breaks some plausible use cases that inexperienced
+users could run into. A user might write the following in a file::
+
+    x: X | None = None
+    class X: ...
+
+Under :pep:`649` this would work fine: ``X`` is not yet defined when it is used,
+but it is not evaluated. However, if a user were to paste this same code into the
+REPL and execute it line by line, it would throw a ``NameError``, because the
+name ``X`` is not yet defined.
```

Comment:
(Somewhat confusing to have a variable named `x` and a class named `X` in the same example)

```suggestion
    a: X | None = None
    class X: ...

Under :pep:`649` this would work fine: ``X`` is not yet defined when it is used
in the annotation for ``a``, but the annotation is lazily evaluated. However,
if a user were to paste this same code into the REPL and execute it line by
line, it would throw a ``NameError``, because the name ``X`` is not yet defined.
```